src/io/command_manager.rs

```diff
@@ -49,37 +90,97 @@ impl CommandManager {
+                // Known issue: if the user changes the rcon_password _in TF2_, this will not trigger an ErrorState change here.
+                ErrorState::Okay | ErrorState::Current(CommandManagerError::Rcon(rcon::Error::Auth)) => {}
+                // Any other issue is worthy of a reconnect attempt.
+                _ => {
+                    match self.try_reconnect().await {
+                        Ok(_) => {
+                            // Current error state (which was _not_ Okay) no presents a historical view on what the error was
+                            // Since we are now connected, if the error state indicates never connected, we can assume first time connect
+                            // Otherwise this is a reconnect.
+                            match self.current_err_state {
+                                ErrorState::Current(_) => {
+                                    tracing::info!("Succesfully reconnected to RCon")
+                                }
+                                ErrorState::Never => {
+                                    tracing::info!("Succesfully established a connection with RCon")
+                                }
+                                _ => {}
+                            };
+                            std::mem::swap(&mut self.current_err_state, &mut self.previous_err_state);
+                            self.current_err_state = ErrorState::Okay;
+                        }
+                        Err(e) => {
+                            // Moves the current error state into the history, and history into current, then override current with the new error.
+                            // This avoids cloning/copying errors by simply moving ownership and dropping scope when not needed.
+                            std::mem::swap(&mut self.current_err_state, &mut self.previous_err_state);
+                            self.current_err_state = ErrorState::Current(e);
+                        }
+                    }
+                }
+            }
+
+            match self
+                .request_recv
+                .recv()
+                .await
+                .expect("The main IO Loop experienced a fatal error.")
+            {
                 CommandManagerMessage::RunCommand(cmd) => {
-                    let cmd = format!("{}", cmd);
-                    if let Err(e) = self.run_command(&cmd).await {
-                        tracing::error!("Failed to run command {}: {:?}", cmd, e);
+                    // Only attempt to run commands if the error state indicates we have a valid RCon client.
+                    // This prevents getting shunted by the TF2 client for repeated Auth failures
+                    if self.current_err_state == ErrorState::Okay {
+                        let cmd = format!("{}", cmd);
+                        if let Err(e) = self.run_command(&cmd).await {
+                            tracing::error!("Failed to run command {}: {:?}", cmd, e);
```

Comment:
Errors are logged here when trying to run the command,and also logged at the top of the loop, resulting in some errors being logged twice. 
![image](https://github.com/MegaAntiCheat/client-backend/assets/47521168/097a0379-5b2a-465c-a3d8-b9f8b8f7ba56)

Reply:
I wonder if maybe the error history should be reserved for the errors that occur when trying to reconnect, and we instead check if it's connected before actually running the requested command. Although just removing this error log would fix the issue I guess.

Hmm I'm not sure I'm 100% happy with the structure of the code for the command manager, to no fault of yours, but refactoring it probably wouldn't achieve anything. If it works.